src/main/java/com/google/sps/servlets/UploadServlet.java

```diff
@@ -0,0 +1,150 @@
+public class UploadServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
+    //  Get all the Dataset Entities
+    Query query = new Query("MetaData").addSort("timestamp", SortDirection.ASCENDING);
+
+    /* Get current user logged in to webapp */
+    UserService userService = UserServiceFactory.getUserService();
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    /*  Get metadata (for comparison and eventual storage) */
+    String userEmail = userService.getCurrentUser().getEmail();
+    String datasetName = request.getParameter("dataset-namer");
+    long timestamp = System.currentTimeMillis();
+
+    // Loop to see if that user already exists
+    boolean datasetPresent = false;
+    boolean userPresent = false;
+    boolean newDataset = true;  // Assume this is a new dataset being created
+    for (Entity entity : results.asIterable()) {
+      String email = (String) entity.getProperty("user-email");
+      String dataset = (String) entity.getProperty("dataset-name");
+
+      if (email.equals(userEmail)) {
+        userPresent = true;
+        if (dataset.equals(datasetName)) {
+          // Do nothing, use this existing dataset
+          datasetPresent = true;
+          newDataset = false;
+        }
+      }
+    }
+    
+    // If this user already has that dataset present
+    if (userPresent && datasetPresent) {
+      response.sendRedirect("/invalid.html");
+    }
+
+
+    if (!datasetPresent) {
+      /* Upload images from form to GCP bucket */
+
+      int imageCount = 0;
+
+      // The ID of your GCP project
+      String projectId = "step-2020-johndallard";
+
+      // The ID of your GCS bucket
+      String bucketName = "embedding-visualizer-bucket";
+
+      Storage storage = StorageOptions.newBuilder().setProjectId(projectId).build().getService();
+
+      // Create new directory within user's subdirectory for this dataset
+      String userDirectory = userEmail + "/";
+      String newDatasetDir = userDirectory + datasetName + "/";
+      String userImagesDir = newDatasetDir + "original_images/";
+      String userThumbnailsDir = newDatasetDir + "thumbnails/";
+      String userSpritesheetDir = newDatasetDir + "spritesheets/";
+      String userEmbeddingsDir = newDatasetDir + "embeddings/";
+      String[] directories = {userImagesDir, userThumbnailsDir, userSpritesheetDir, userEmbeddingsDir};
+
+      BlobId blobId = BlobId.of(bucketName, userDirectory);
+      BlobInfo blobInfo = BlobInfo.newBuilder(blobId).build();
+      storage.create(blobInfo);
+
+      // Create new subdirectories for the new dataset
+      for (int i = 0; i < directories.length; i++) {
+        blobId = BlobId.of(bucketName, directories[i]);
+        blobInfo = BlobInfo.newBuilder(blobId).build();
+        storage.create(blobInfo);
```

Comment:
hmm, just my two cents. never realized this until now, but I actually have to create the folders in the bucket that I upload all my results to (just how blob uploading works in python). No need to delete them, but just lyk

Reply:
Ohh I see. So what your saying is I should only create the original_images/ directory, and leave the creation of the other three directories (embeddings/, spritesheets/, and thumbnails/) to you?

Reply:
I don't think that's true-- the directories create themselves when you include them in the blob filename. But it _is_ true that you don't have to do all that here, Myles.

Reply:
I ended up removing those three directories from datasets when they're created, but it could always put them back if needed. (This part is very flexible)